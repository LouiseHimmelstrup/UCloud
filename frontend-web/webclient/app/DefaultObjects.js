import {Cloud} from "../authentication/SDUCloudObject"
import {tusConfig} from "./Configurations";
import Uppy from "uppy";

export const DefaultStatus = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended.",
};

export const RightsMap = {
    NONE: 0,
    READ: 1,
    READ_WRITE: 2,
    OWN: 3,
};

export const AnalysesStatusMap = {
    "PENDING": 0,
    "IN PROGRESS": 1,
    "COMPLETED": 2,
};

export const RightsNameMap = {
    NONE: "None",
    READ: "Read",
    READ_WRITE: "Read/Write",
    OWN: "Own",
};

export const SensitivityLevel = {
    "OPEN_ACCESS": "Open Access",
    "CONFIDENTIAL": "Confidential",
    "SENSITIVE": "Sensitive",
};

export const SensitivityLevelMap = {
    "OPEN_ACCESS": 0,
    "CONFIDENTIAL": 1,
    "SENSITIVE": 2,
};

const initializeUppy = (restrictions) => 
    Uppy.Core({
        autoProceed: false,
        debug: false,
        restrictions: restrictions,
        meta: {
            sensitive: false,
        },
        onBeforeUpload: () => {
            return Cloud.receiveAccessTokenOrRefreshIt().then((data) => {
                tusConfig.headers["Authorization"] = `Bearer ${data}`;
            });
        }
    }).use(Uppy.Tus, tusConfig);

export const initObject = {
    files: {
        files: [],
        filesShown: 10,
        filesPerPage: 10,
        currentFilesPage: 0,
        loading: false,
        path: "",
        filesInfoPath: "",
        filesInfoFile: null,
        projects: []
    },
    uppy: {
        uppyFiles: initializeUppy({maxNumberOfFiles: false}),
        uppyFilesOpen: false,
        uppyRunApp: initializeUppy({ maxNumberOfFiles: 1}),
        uppyRunAppOpen: false
    },
    status: {
        status: DefaultStatus,
        title: ""
    }
};