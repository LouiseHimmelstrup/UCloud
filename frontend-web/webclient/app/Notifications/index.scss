.ui.feed > .event.notification {
    padding: 5px;
    border-radius: 5px;
    cursor: pointer;
    width: 350px;
}

.notification-trigger {
    cursor: pointer;
}

.ui.feed > .event.notification.unread {
    background-color: #ddd;
}

.ui.feed > .event.notification.unread:hover, .notification.read:hover {
    background-color: #eee;
}

button.ui.blue.circular.notification-button-padding, i.notification-button-padding {
    padding-left: 8px;
    padding-right: 8px;
}