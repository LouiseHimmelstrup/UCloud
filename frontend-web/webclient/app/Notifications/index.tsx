import * as React from "react";
import { Cloud } from "../../authentication/SDUCloudObject"
import { Popup, Feed, Icon, SemanticICONS, Label, Button, Divider } from 'semantic-ui-react';
import { Redirect } from "react-router";
import * as moment from "moment";
import { connect } from "react-redux";
import "./index.scss";
import { withRouter } from "react-router";
import { Page } from "../Types";
import { fetchNotifications, notificationRead } from "./Redux/NotificationsActions";
import { History } from "history";
import Status from "../Navigation/Status";

interface NotificationProps {
    page: Page<Notification>
    redirectTo: string
    fetchNotifications: Function,
    notificationRead: Function,
    history: History
}

class Notifications extends React.Component<NotificationProps> {
    constructor(props) {
        super(props);
    }

    public componentDidMount() {
        this.reload();
        setInterval(() => this.reload(), 30_000);
    }

    private reload() {
        this.props.fetchNotifications();
    }

    private onNotificationRead(notification: Notification) {
        // TODO This is not the "react" way
        this.props.notificationRead(notification.id);
        Cloud.post(`notifications/read/${notification.id}`)
    }

    private onNotificationAction(notification: Notification) {
        switch (notification.type) {
            case "APP_COMPLETE":
                // TODO This is buggy! Does't update if already present on analyses page
                // TODO Should refactor these URLs somewhere else
                this.props.history.push(`/analyses/${notification.meta.jobId}`);
                break;
            case "SHARE_REQUEST":
                // TODO This is a bit lazy
                this.props.history.push("/shares");
                break;
        }
    }

    public render() {
        let entries: JSX.Element[] = this.props.page.items.map((notification, index) =>
            <NotificationEntry
                key={index}
                notification={notification}
                onMarkAsRead={(it) => this.onNotificationRead(it)}
                onAction={(it) => this.onNotificationAction(it)}
            />
        );

        if (this.props.redirectTo) {
            let theRedirect = this.props.redirectTo;
            return <Redirect to={theRedirect} />
        }

        const unreadLength = this.props.page.items.filter((e) => !e.read).length;
        const label = unreadLength > 0 ? (
            <Label color='red' floating circular>
                {unreadLength}
            </Label>
        ) : null;
        return (
            <Popup
                trigger={
                    <Button color='blue' className="notification-button-padding" circular>
                        <Icon className="notification-button-padding" name="bell" />
                        {label}
                    </Button>
                }
                content={
                    <Feed>
                        {entries.length ? entries : <NoNotifications />}
                        <Divider />
                        <Status />
                    </Feed>}
                on="click"
                position="bottom right"
            />
        );
    }
}

const NoNotifications = () =>
    <Feed.Event className="notification">
        <Feed.Content>
            <Feed.Label>No notifications</Feed.Label>
        </Feed.Content>
    </Feed.Event>

export interface Notification {
    type: string
    id: any
    message: string
    ts: Number
    read: boolean
    meta: any
}

interface NotificationEntryProps {
    notification: Notification
    onMarkAsRead?: (notification: Notification) => void
    onAction?: (notification: Notification) => void
}

class NotificationEntry extends React.Component<NotificationEntryProps, any> {
    constructor(props: NotificationEntryProps) {
        super(props);
    }

    public render() {
        return (
            <Feed.Event className={"notification " + (this.props.notification.read ? "read " : "unread ")} onClick={() => this.handleAction()}>
                <Feed.Label><Icon name={this.resolveEventIcon(this.props.notification.type)} /></Feed.Label>
                <Feed.Content>
                    <Feed.Date content={moment(this.props.notification.ts.toString(), "x").fromNow()} />
                    <Feed.Summary>{this.props.notification.message}</Feed.Summary>
                </Feed.Content>
            </Feed.Event>
        );
    }

    private handleRead() {
        if (this.props.onMarkAsRead) this.props.onMarkAsRead(this.props.notification);
    }

    private handleAction() {
        this.handleRead();
        if (this.props.onAction) this.props.onAction(this.props.notification);
    }

    private resolveEventIcon(eventType: string): SemanticICONS {
        switch (eventType) {
            case "APP_COMPLETE": return "tasks";
            case "SHARE_REQUEST": return "share";
            default: return "question";
        }
    }
}

const mapDispatchToProps = (dispatch) => ({
    fetchNotifications: () => dispatch(fetchNotifications()),
    notificationRead: (id) => dispatch(notificationRead(id))
});
const mapStateToProps = (state) => state.notifications;

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Notifications));