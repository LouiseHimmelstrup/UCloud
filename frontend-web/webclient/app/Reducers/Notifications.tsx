export const RECEIVE_NOTIFICATIONS = "RECEIVE_NOTIFICATIONS";
export const NOTIFICATION_READ = "NOTIFICATION_READ";
export const SET_REDIRECT = "SET_REDIRECT";

const Notifications = (state: any = {}, action: any) => {
    switch (action.type) {
        case RECEIVE_NOTIFICATIONS: {
            return { ...state, page: action.page };
        }
        case NOTIFICATION_READ: {
            return {
                ...state, items: state.page.items.map((n) => {
                    if (n.id === action.id) n.read = true;
                    return n;
                })
            }
        }
        case SET_REDIRECT: {
            return { ...state, redirectTo: action.redirectTo }
        }
        default: {
            return state;
        }
    }
}

export default Notifications;