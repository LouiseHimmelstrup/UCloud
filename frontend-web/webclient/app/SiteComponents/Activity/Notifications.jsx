import React from 'react';
import $ from 'jquery';
import LoadingIcon from '../LoadingIcon'
import { NotificationIcon, WebSocketSupport } from '../../UtilityFunctions'
import pubsub from "pubsub-js";

class Notifications extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            recent: [],
            remaining: [],
            loading: true,
            hasWebSocketSupport: "WebSocket" in window,
            currentNotification: {
                body: "",
                timestamp: 0,
                message: "",
            },
            // FIXME: Should be WSS
            ws: new WebSocket("ws://localhost:8080/ws/notifications"),
            recentShown: 10,
            remainingShown: 10,
        }
    }

    componentDidMount() {
        this.getNotifications();
        pubsub.publish('setPageTitle', this.constructor.name);
    }

    getNotifications() {
        //this.setState({loading: true});
        /*$.getJSON("http://localhost:8080/api/getNotifications").then((notifications) => {
            let yesterday = new Date().getTime() - 24 * 60 * 60 * 1000;
            let recentNotifications = this.state.recent.slice();
            let remainingNotifications = this.state.remaining.slice();
            notifications.forEach((it) => {
                if (it.timestamp > yesterday) {
                    recentNotifications.push(it);
                } else {
                    remainingNotifications.push(it);
                }
            });
            recentNotifications.sort();
            remainingNotifications.sort();
            this.setState({
                loading: false,
                recent: recentNotifications,
                remaining: remainingNotifications,
            });
            //if (this.state.hasWebSocketSupport) this.initWS();
        });*/
    }

    initWS() {
        this.state.ws.onerror = () => {
            console.log("Socket error.")
        };

        this.state.ws.onopen = () => {
            console.log("Connected");
        };

        this.state.ws.onmessage = response => {
            let recentList = this.state.recent.slice();
            recentList.push(JSON.parse(response.data));
            this.setState({
                recent: recentList,
            });
        };
    }

    updateCurrentNotification(notification) {
        this.setState({
            currentNotification: notification
        })
    }

    showMore(name) {
        if (name === "recent") {
            this.setState({
                recentShown: this.state.recentShown + 10
            });
        } else {
            this.setState({
                remainingShown: this.state.remainingShown + 10
            });
        }
    }

    render() {
        return (
            <section>
                <div className="container container-md">
                    <LoadingIcon loading={this.state.loading}/>
                    <WebSocketSupport/>
                    <p className="ph">Last 24 hours</p>
                    <div className="card">
                        <table className="table table-hover table-fixed va-middle">
                            <NotificationList onClick={(notification) => this.updateCurrentNotification(notification)} notifications={this.state.recent} showCount={this.state.recentShown}/>
                        </table>
                        <ShowButton onClick={() => this.showMore("recent")} hasMoreNotifications={this.state.recent.length > this.state.recentShown}/>
                    </div>
                    <p className="ph">Older</p>
                    <div className="card">
                        <table className="table table-hover table-fixed va-middle">
                            <NotificationList onClick={(notification) => this.updateCurrentNotification(notification)} notifications={this.state.remaining} showCount={this.state.remainingShown}/>
                        </table>
                        <ShowButton onClick={() => this.showMore("remaining")} hasMoreNotifications={this.state.remaining.length > this.state.remainingShown}/>
                    </div>
                </div>
                <MessageModal notification={this.state.currentNotification}/>
            </section>)
    }
}

function NotificationList(props) {
    if (!props.notifications) return (null);
    const notifications = props.notifications.slice(0, Math.min(props.showCount, props.notifications.length));
    let i = 0;
    const notificationsList = notifications.map((notification) =>
        <tr key={i++} className="msg-display clickable" onClick={() => props.onClick(notification)} data-toggle="modal"
            data-target="#notificationModal">
            <td className="wd-xxs">
                <NotificationIcon type={notification.type}/>
            </td>
            <th className="mda-list-item-text mda-2-line">
                <small>{notification.message}</small>
                <br/>
                <small className="text-muted">{new Date(notification.timestamp).toLocaleString()}</small>
            </th>
            <td className="text">{notification.body}</td>
        </tr>
    );

    return (
        <tbody>
        {notificationsList}
        </tbody>)
}

function ShowButton(props) {
    if (props.hasMoreNotifications) {
        return (
            <button onClick={() => props.onClick()} className="btn btn-info ion-ios-arrow-down"/>)
    } else {
        return (<div/>)
    }
}

function MessageModal(props) {
    return (
        <div id="notificationModal" className="modal fade" role="dialog">
            <div className="modal-dialog">
                <div className="modal-content">
                    <div className="modal-header">
                        <button type="button" className="close" data-dismiss="modal">&times;</button>
                        <h4 className="modal-title"> {props.notification.message}<br/>
                            <small>{ new Date(props.notification.timestamp).toLocaleString() }</small>
                        </h4>
                    </div>
                    <div className="modal-body">
                        <p>{props.notification.body}</p>
                    </div>
                    <div className="modal-footer">
                        <button type="button" className="btn btn-default" data-dismiss="modal">Close</button>
                    </div>
                </div>
            </div>
        </div>)
}

export default Notifications
