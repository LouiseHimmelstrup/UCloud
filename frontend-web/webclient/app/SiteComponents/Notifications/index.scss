.notification {
    padding: 5px !important;
    border-radius: 5px;
    cursor: pointer;
    width: 350px !important;
}

.notification.unread {
    background-color: #ddd !important;
}