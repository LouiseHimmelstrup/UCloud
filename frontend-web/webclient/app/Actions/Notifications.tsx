import { RECEIVE_NOTIFICATIONS, NOTIFICATION_READ, SET_REDIRECT } from "../Reducers/Notifications";
import { Cloud } from "../../authentication/SDUCloudObject";
import { Page } from "../types/types";
import { failureNotification } from "../UtilityFunctions";

const ERROR = "ERROR";

type Action = { type: string }
interface ReceiveAction extends Action { page: Page<Notification> }
interface ReadAction extends Action { id: Number }

const receiveNotifications = (page: Page<Notification>): ReceiveAction => ({
    type: RECEIVE_NOTIFICATIONS,
    page
})

export const fetchNotifications = () =>
    Cloud.get("/notifications")
        .then(({ response }: { response: Page<Notification> }) => receiveNotifications(response))
        .catch(() => {
            failureNotification("Failed to retrieve notifications, please try again later");
            return NoAction();
        });;

export const notificationRead = (id: Number): ReadAction => ({
    type: NOTIFICATION_READ,
    id
});

export const setRedirectTo = (redirectTo: string) => ({
    type: SET_REDIRECT,
    redirectTo
});

const NoAction = () => ({ type: ERROR })