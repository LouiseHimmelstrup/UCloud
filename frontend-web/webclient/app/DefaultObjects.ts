import {SidebarOption, Page} from "Types";
import {Status} from "Navigation";
import {Analysis, DetailedApplicationSearchReduxState, RunsSortBy} from "Applications";
import {File, DetailedFileSearchReduxState} from "Files";
import {SortOrder, SortBy} from "Files";
import {DashboardStateProps} from "Dashboard";
import {Publication} from "Zenodo";
import {Notification} from "Notifications";
import {Upload} from "Uploader";
import {Activity, ActivityGroup, ActivityFilter} from "Activity";
import {Reducer} from "redux";
import {SimpleSearchStateProps} from "Search";
import * as ApplicationRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import * as SnackbarRedux from "Snackbar/Redux";
import * as FavoritesRedux from "Favorites/Redux";
import {defaultAvatar} from "UserSettings/Avataaar";
import {SidebarPages} from "ui-components/Sidebar";
import {ScrollResult} from "Scroll/Types";
import * as ProjectRedux from "Project/Redux";

export const DefaultStatus: Status = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended."
};

export enum KeyCode {
    ENTER = 13,
    ESC = 27
}

export const emptyPage: Page<any> = {items: [], itemsPerPage: 25, itemsInTotal: 0, pageNumber: 0, pagesInTotal: 0};

export enum AnalysesStatusMap {
    "PENDING",
    "IN PROGRESS",
    "COMPLETED"
}

export enum RightsNameMap {
    "NONE" = "None",
    "READ" = "Read",
    "READ_WRITE" = "Read/Write"
}

export enum SensitivityLevel {
    "INHERIT" = "Inherit",
    "PRIVATE" = "Private",
    "CONFIDENTIAL" = "Confidential",
    "SENSITIVE" = "Sensitive"
}

export type Sensitivity = keyof typeof SensitivityLevel;

export enum SensitivityLevelMap {
    INHERIT = "INHERIT",
    PRIVATE = "PRIVATE",
    CONFIDENTIAL = "CONFIDENTIAL",
    SENSITIVE = "SENSITIVE"
}

function getFilesSortingColumnOrDefault(columnIndex: 0 | 1): SortBy {
    const sortingColumn = window.localStorage.getItem(`filesSorting${columnIndex}`);
    if (sortingColumn && Object.values(SortBy).includes(sortingColumn)) return sortingColumn as SortBy;
    switch (columnIndex) {
        case 0:
            window.localStorage.setItem("filesSorting0", SortBy.MODIFIED_AT);
            return SortBy.MODIFIED_AT;
        case 1:
            window.localStorage.setItem("filesSorting1", SortBy.SIZE);
            return SortBy.SIZE;
    }
}

function getItemOrDefault<T, T2>(itemName: string, defaultValue: T, en: T2): T {
    const item = window.localStorage.getItem(itemName);
    if (item && Object.values(en).includes(item)) {
        return item as unknown as T;
    }
    return defaultValue;
}

export interface ComponentWithLoadingState {
    loading: boolean
    error?: string
}

export interface ComponentWithPage<T> extends ComponentWithLoadingState {
    page: Page<T>
}

export interface ComponentWithScroll<Item, OffsetType> extends ComponentWithLoadingState {
    scroll?: ScrollResult<Item, OffsetType>
}

export interface ResponsiveReduxObject {
    mediaType: string
    orientation: string
    lessThan: Record<string, boolean>
    greaterThan: Record<string, boolean>
    is: Record<string, boolean>
}

export const initResponsive = (): ResponsiveReduxObject => ({
    mediaType: "",
    orientation: "",
    lessThan: {},
    greaterThan: {},
    is: {},
});

export interface FilePreviewReduxState {
    file?: File
    error?: string
}

export interface FilesReduxObject extends ComponentWithPage<File> {
    sortOrder: SortOrder
    sortBy: SortBy
    path: string
    filesInfoPath: string
    fileSelectorError?: string
    sortingColumns: [SortBy, SortBy]
    fileSelectorLoading: boolean
    fileSelectorShown: boolean
    fileSelectorPage: Page<File>
    fileSelectorPath: string
    fileSelectorCallback: Function
    fileSelectorIsFavorites: boolean
    disallowedPaths: string[]
    invalidPath: boolean
}

export interface FileInfoReduxObject {
    file?: File
    error?: string
    activity: Page<Activity>
    loading: boolean
}

export interface AnalysisReduxObject extends ComponentWithPage<Analysis> {
    sortBy: RunsSortBy
    sortOrder: SortOrder
}

export interface NotificationsReduxObject {
    redirectTo: string
    items: Notification[]
    loading: boolean
    error?: string
}

export interface ZenodoReduxObject extends ComponentWithPage<Publication> {
    connected: boolean
}

export interface StatusReduxObject {
    status: Status
    title: string
    page: SidebarPages
    loading: boolean
}

export interface SidebarReduxObject {
    pp: boolean
    options: SidebarOption[]
    kcCount: number
}

export interface HeaderSearchReduxObject {
    prioritizedSearch: HeaderSearchType
    refresh?: () => void
}

export interface RunApplicationReduxObject {

}

export type ActivityReduxObject = ComponentWithScroll<ActivityGroup, number> & ActivityFilter;

export type HeaderSearchType = "files" | "applications" | "projects";

export interface UploaderReduxObject {
    uploads: Upload[]
    visible: boolean
    path: string
    allowMultiple: boolean
    onFilesUploaded: (p: string) => void
    error?: string
    loading: boolean
}

interface LegacyReducers {
    dashboard?: Reducer<DashboardStateProps>
    files?: Reducer<FilesReduxObject>
    uploader?: Reducer<UploaderReduxObject>
    status?: Reducer<StatusReduxObject>
    notifications?: Reducer<NotificationsReduxObject>
    analyses?: Reducer<AnalysisReduxObject>
    zenodo?: Reducer<ZenodoReduxObject>
    header?: Reducer<HeaderSearchReduxObject>
    sidebar?: Reducer<SidebarReduxObject>
    activity?: Reducer<ActivityReduxObject>
    detailedResult?: Reducer<DetailedResultReduxObject>
}

export type Reducers = LegacyReducers & ApplicationRedux.Reducers & AccountingRedux.Reducers;

export type DetailedResultReduxObject = ComponentWithPage<File>

export const initDetailedResult = (): DetailedResultReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined
});

/* FIXME */
interface LegacyReduxObject {
    dashboard: DashboardStateProps
    files: FilesReduxObject,
    uploader: UploaderReduxObject
    status: StatusReduxObject,
    notifications: NotificationsReduxObject
    analyses: AnalysisReduxObject
    zenodo: ZenodoReduxObject
    header: HeaderSearchReduxObject
    sidebar: SidebarReduxObject
    activity: ActivityReduxObject
    detailedResult: DetailedResultReduxObject
    simpleSearch: SimpleSearchStateProps
    detailedFileSearch: DetailedFileSearchReduxState
    detailedApplicationSearch: DetailedApplicationSearchReduxState
    fileInfo: FileInfoReduxObject
    avatar: AvatarReduxObject
    filePreview: FilePreviewReduxState
    responsive?: ResponsiveReduxObject
    project: ProjectRedux.State
    loading?: boolean
}

export type ReduxObject =
    LegacyReduxObject &
    ApplicationRedux.Objects &
    AccountingRedux.Objects &
    FavoritesRedux.Objects &
    SnackbarRedux.Wrapper;

export const initActivity = (): ActivityReduxObject => ({
    loading: false
});

export const initNotifications = (): NotificationsReduxObject => ({
    items: [],
    loading: false,
    redirectTo: "",
    error: undefined
});

export const initHeader = (): HeaderSearchReduxObject => ({
    prioritizedSearch: "files"
});

export const initStatus = (): StatusReduxObject => ({
    status: DefaultStatus,
    title: "",
    page: SidebarPages.None,
    loading: false
});

export const initDashboard = (): DashboardStateProps => ({
    favoriteFiles: [],
    recentFiles: [],
    recentAnalyses: [],
    notifications: [],
    favoriteLoading: false,
    recentLoading: false,
    analysesLoading: false
});

export function initObject(homeFolder: string): ReduxObject {
    return {
        dashboard: initDashboard(),
        files: initFiles(homeFolder),
        status: initStatus(),
        header: initHeader(),
        notifications: initNotifications(),
        analyses: initAnalyses(),
        zenodo: initZenodo(),
        sidebar: initSidebar(),
        uploader: initUploads(),
        activity: initActivity(),
        detailedResult: initDetailedResult(),
        simpleSearch: initSimpleSearch(),
        detailedApplicationSearch: initApplicationsAdvancedSearch(),
        detailedFileSearch: initFilesDetailedSearch(),
        fileInfo: initFileInfo(),
        avatar: initAvatar(),
        filePreview: initFilePreview(),
        project: ProjectRedux.initialState,
        ...ApplicationRedux.init(),
        ...AccountingRedux.init(),
        ...FavoritesRedux.init(),
        ...SnackbarRedux.init(),
        responsive: undefined,
    }
}


export const initFilePreview = () => ({
    file: undefined,
    error: undefined
});

export type AvatarReduxObject = typeof defaultAvatar & { error?: string };
export const initAvatar = (): AvatarReduxObject => ({...defaultAvatar, error: undefined});

export const initSimpleSearch = (): SimpleSearchStateProps => ({
    files: emptyPage,
    filesLoading: false,
    applications: emptyPage,
    applicationsLoading: false,
    errors: [],
    search: "",
    applicationSearch: initApplicationsAdvancedSearch(),
    fileSearch: initFilesDetailedSearch()
});

export const initAnalyses = (): AnalysisReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined,
    sortBy: RunsSortBy.createdAt,
    sortOrder: SortOrder.ASCENDING
});

export const initZenodo = (): ZenodoReduxObject => ({
    connected: false,
    loading: false,
    page: emptyPage
});

export const initSidebar = (): SidebarReduxObject => ({
    pp: false,
    kcCount: 0,
    options: []
});

export const initUploads = (): UploaderReduxObject => ({
    path: "",
    uploads: [],
    visible: false,
    allowMultiple: false,
    error: undefined,
    onFilesUploaded: () => null,
    loading: false
});

export const initFileInfo = (): FileInfoReduxObject => ({
    activity: emptyPage,
    loading: false
});

export const initFiles = (homeFolder: string): FilesReduxObject => ({
    page: emptyPage,
    sortOrder: getItemOrDefault("sortOrder", SortOrder.ASCENDING, SortOrder),
    sortBy: getItemOrDefault("sortBy", SortBy.PATH, SortBy),
    loading: false,
    error: undefined,
    path: "",
    filesInfoPath: "",
    sortingColumns: [getFilesSortingColumnOrDefault(0), getFilesSortingColumnOrDefault(1)],
    fileSelectorLoading: false,
    fileSelectorShown: false,
    fileSelectorPage: emptyPage,
    fileSelectorPath: homeFolder,
    fileSelectorIsFavorites: false,
    fileSelectorCallback: () => undefined,
    fileSelectorError: undefined,
    disallowedPaths: [],
    invalidPath: false
});

export const initFilesDetailedSearch = (): DetailedFileSearchReduxState => ({
    hidden: true,
    allowFolders: true,
    allowFiles: true,
    fileName: "",
    extensions: new Set(),
    tags: new Set(),
    sensitivities: new Set(),
    createdBefore: undefined,
    createdAfter: undefined,
    modifiedBefore: undefined,
    modifiedAfter: undefined,
    error: undefined,
    loading: false
});

export const initApplicationsAdvancedSearch = (): DetailedApplicationSearchReduxState => ({
    error: undefined,
    loading: false,
    hidden: true,
    appName: "",
    appVersion: "",
    tags: ""
});
