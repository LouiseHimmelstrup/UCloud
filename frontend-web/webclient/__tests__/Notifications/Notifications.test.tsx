import Notifications from "Notifications/index";
import { create } from "react-test-renderer";
import * as React from "react";
import { Provider } from "react-redux";
import { configureStore } from "Utilities/ReduxUtilities";
import { initNotifications, initUploads } from "DefaultObjects";
import notifications from "Notifications/Redux/NotificationsReducer";
import uploader from "Uploader/Redux/UploaderReducer";
import { MemoryRouter } from "react-router";

describe.skip("Notifications", () => {
    // FIXME Will try to contact backend and get wrong result, overwriting the page
    test("Mount notifications", () => {
        expect(create(
            <Provider store={configureStore({ notifications: initNotifications(), uploader: initUploads() }, { notifications, uploader })}>
                <MemoryRouter>
                    <Notifications />
                </MemoryRouter>
            </Provider>
        )).toMatchSnapshot();
    })
}); 