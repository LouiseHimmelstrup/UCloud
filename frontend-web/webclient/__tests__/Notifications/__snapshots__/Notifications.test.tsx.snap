// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications Mount notifications 1`] = `
<button
  className="ui blue circular button notification-button-padding"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
>
  <i
    aria-hidden="true"
    className="bell icon notification-button-padding"
  />
</button>
`;
